File: 0101-symmetric-tree/0101-symmetric-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean check(TreeNode left, TreeNode right)
    {
        int leftval = -999;
        int rightval = -999;
        if(left!=null)  leftval = left.val;
        if(right!=null) rightval = right.val;
        System.out.println("Checking " + leftval + " and " + rightval);
        if(left == null && right == null)   return true;
        else if(left == null || right == null)  return false;
        if(left.val != right.val)   return false;
        return check(left.left, right.right) && check(left.right, right.left);
    }
    public boolean isSymmetric(TreeNode root) 
    {
        return check(root, root);
    }
}

File: 0101-symmetric-tree/0101-symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool helper(TreeNode* x, TreeNode* y) {
        if(x == nullptr ^ y == nullptr) return false;
        if(!x && !y) return true;

        if(x->val != y->val) return false;
        return helper(x->left , y->right) && helper(x->right , y->left);
    }

    bool isSymmetric(TreeNode* root) {
        if(!root) return false;
        return helper(root->left, root->right);
    }
};